### ISSUE TITLE
Potential integration with ocaml-css parser?

### ISSUE THREAD

**Issue by jchavarri** [astrada/ppx_bs_css#25]
I've recently learnt about `ocaml-css`, which also includes a css parser: https://github.com/samouri/ocaml-css/blob/master/lib/Parser.mly

I was wondering if it would make sense to join forces and create a single project / repo with just the parser + common AST types? So `ocaml-css` can use it to generate JSON, this PPX can use it too, and also any other project that parses CSS?

cc @samouri @ryyppy

**Comment by astrada** [comment]
Sure, why not? If @samouri is interested, I'm in.

I'm using sedlex instead of ocamllex, but that should not be a problem. And I'm specifically targeting CSS3 to simplify things.

**Comment by jchavarri** [comment]
@astrada That sounds great! I might be able to help with this. Maybe I can start by creating a new repo and trying to extricate the parser parts from this one, leveraging esy as much as possible. I think it would be fair to transfer the ownership of that new repo afterwards. I'd feel more comfortable that way, as it would still be 99.9% your code 😄 I'm open to any option.

One q, out of curiosity, do you know what are the main differences between sedlex and ocamllex? From what I've seen, the main advantage of sedlex is the support for unicode, but wonder if there are other trade-offs.

> regarding ocaml-css and apparently this tool was intended for fastpack

@ryyppy What do you mean with this? What are the technical aspects that make the parser more friendly towards fastpack?

**Comment by ryyppy** [comment]
Just saying that this project apparently raised from the idea of parsing CSS natively. Fastpack relies on NodeJS tools to do the css processing afaik. It's not relevant for the ppx, just saying that we have many intersection points where this parser lib could be useful!

**Comment by astrada** [comment]
OK, here you go: https://github.com/astrada/ocaml-css-parser

**Comment by samouri** [comment]
Wow I had no idea there was another css parser written in ocaml floating around!

@ryyppy is right -- I initially started working on [ocaml-css](https://github.com/samouri/ocaml-css) for the purposes of integrating with FastPack. I haven't had as much time to work on it recently and never got it to 100% complete. The way I've been measuring progress is that I took all of the test-cases from [https://github.com/reworkcss/css](reworkcss/css) and have working on them one at a time. I believe I'm currently at ~40%.  It took me awhile to figure out how to support comments 😬, but now that I've completed that, I believe the remaining test cases are relatively straightforward.

> I'm using sedlex instead of ocamllex,

I had been planning on switching from ocamllex to `Ulex`  (or `sedlex`) for the utf-8 support but haven't had a chance to yet. I'm also not sure of what the difference is between the two. 

>If we could get a parser (i don't care which one, as long as it's a plain parser library) to a point where it's covering most use-cases, that would be a real asset for the Reason / OCaml community!

Couldn't agree more. For my part I'll try do evaluate how far away my lib is from being 100% complete.

@astrada: does this parser support:

1. comments in the ast
2. printing back out to css
3. error detection. If yes detecting multiple errors or just one and then fail fast?
4. compile to js via either jsoo or bucklescript
